### PR TITLE
Adding ConfigMap build inputs

### PR DESCRIPTION
**Pull request by adambkaplan**
Re-worked example to use Java/Maven application (which is amenable to both ConfigMap and Secret data).

Trello card: https://trello.com/c/RMKJxJUm/1020-5-allow-using-a-configmap-as-an-input-to-a-build-builds

### CODE REVIEW

**Comment by adambkaplan** [comment]
/hold until 3.10 branch is cut
/assign @bmcelvee 

ptal at style - when to use "config map" vs. "Config Map" vs. `ConfigMap`

**Comment by bparees** [comment]
i've added the `online` release tag but technically this won't be in online until online is update to 3.11.  Not sure if there's a more appropriate way to handle that.

File: dev_guide/builds/build_inputs.adoc [hunk 1]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
-During the build, the *_.npmrc_* file is copied into the directory where the
-source code is located. In {product-title} S2I builder images, this
-is the image working directory, which is set using the `WORKDIR` instruction
-in the *_Dockerfile_*. If you want to specify another directory, add a
-`destinationDir` to the secret definition:
+During the build, the *_settings.xml_* and *_id_rsa_* files are copied into the
+directory where the source code is located. In {product-title} S2I builder 
+images, this is the image working directory, which is set using the `WORKDIR` 
+instruction in the *_Dockerfile_*. If you want to specify another directory, 
+add a `destinationDir` to the secret definition:
```
Comment (bparees) [inline dev_guide/builds/build_inputs.adoc]: s/secret//

File: dev_guide/builds/build_inputs.adoc [hunk 2]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
 You can also specify the destination directory when creating a new
 `BuildConfig`:
 
 ----
 $ oc new-build \
-    openshift/nodejs-010-centos7~https://github.com/openshift/nodejs-ex.git \
-    --build-secret “secret-npmrc:/etc”
+    openshift/wildfly-101-centos7~https://github.com/wildfly/quickstart.git \
+    --contextDir helloworld --build-secret “secret-mvn:.ssh” \
+    --build-configMap "settings-mvn:.m2"
```
Comment (bparees) [inline dev_guide/builds/build_inputs.adoc]: looks like it's `--build-config-map`? 
https://github.com/openshift/origin/pull/19655/files#diff-924f9baa301d1772b396b1629eb2bb15R118

File: dev_guide/builds/build_inputs.adoc [hunk 3]
```diff
@@ -830,10 +859,12 @@ created during the copy process.
 
 [NOTE]
 ====
-Currently, any files with these secrets are world-writable (have `0666`
-permissions) and will be truncated to size zero after executing the *_assemble_*
-script. This means that the secret files will exist in the resulting image, but
-they will be empty for security reasons.
+Input secrets are added with world-writable (have `0666` permissions) and will
```
Comment (bparees) [inline dev_guide/builds/build_inputs.adoc]: s/with/as/

File: dev_guide/builds/build_inputs.adoc [hunk 4]
```diff
@@ -853,22 +884,22 @@ into that directory, relative to your *_Dockerfile_* location. This makes the
...
-.Example of a Dockerfile referencing secret data
+.Example of a Dockerfile referencing secret and configMap data
 ====
 ----
 FROM centos/ruby-22-centos7
 
 USER root
 ADD ./secret-dir /secrets
-COPY ./secret2 /
+COPY ./config /
```
Comment (bparees) [inline dev_guide/builds/build_inputs.adoc]: not sure why one of these is doing ADD and the other COPY, let's make them consistent.

**Review by bparees — CHANGES_REQUESTED** [review]

**Comment by adambkaplan** [comment]
[(5) Allow using a ConfigMap as an input to a build [builds]](https://trello.com/c/RMKJxJUm/1020-5-allow-using-a-configmap-as-an-input-to-a-build-builds)

File: dev_guide/builds/build_inputs.adoc [hunk 5]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
-=== Adding Input Secrets
+=== Adding Input Secrets and ConfigMaps
 
-To add an input secret to an existing `BuildConfig`:
+To add an input secret and/or config map to an existing `BuildConfig`:
+
+. Create the config map, if it does not exist:
++
+----
+$ oc create configMap settings-mvn \
```
Comment (xiuwang) [inline dev_guide/builds/build_inputs.adoc]: s/configMap/configmap

File: dev_guide/builds/build_inputs.adoc [hunk 6]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
-To include the secret in a new `BuildConfig`, run the following command:
+To include the secret and configMap in a new `BuildConfig`, run the following 
+command:
 
 ----
 $ oc new-build \
-    openshift/nodejs-010-centos7~https://github.com/openshift/nodejs-ex.git \
-    --build-secret secret-npmrc
+    openshift/wildfly-101-centos7~https://github.com/wildfly/quickstart.git \
+    --contextDir helloworld --build-secret “secret-mvn” \
```
Comment (xiuwang) [inline dev_guide/builds/build_inputs.adoc]: s/--contextDir/--context-dir

File: dev_guide/builds/build_inputs.adoc [hunk 7]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
 
 You can also specify the destination directory when creating a new
 `BuildConfig`:
 
 ----
 $ oc new-build \
-    openshift/nodejs-010-centos7~https://github.com/openshift/nodejs-ex.git \
-    --build-secret “secret-npmrc:/etc”
+    openshift/wildfly-101-centos7~https://github.com/wildfly/quickstart.git \
+    --contextDir helloworld --build-secret “secret-mvn:.ssh” \
```
Comment (xiuwang) [inline dev_guide/builds/build_inputs.adoc]: s/--contextDir/--context-dir

File: dev_guide/builds/build_inputs.adoc [hunk 8]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
+tokens, license files, and more.
 
 [[using-secrets-in-the-buildconfig]]
-=== Adding Input Secrets
+=== Adding Input Secrets and ConfigMaps
 
-To add an input secret to an existing `BuildConfig`:
+To add an input secret and/or config map to an existing `BuildConfig`:
+
+. Create the config map, if it does not exist:
```
Comment (xiuwang) [inline dev_guide/builds/build_inputs.adoc]: s/config map/configmap ?

Comment (adambkaplan) [reply dev_guide/builds/build_inputs.adoc]: I'll defer to @bmcelvee on conventions around kube objects like `ConfigMap`

Comment (bmcelvee) [reply dev_guide/builds/build_inputs.adoc]: Let's go with ConfigMap.

File: dev_guide/builds/build_inputs.adoc [hunk 9]
```diff
@@ -23,7 +23,7 @@ endif::[]
 * xref:image-source[Content extracted from existing images]
 * xref:source-code[Git repositories]
 * xref:binary-source[Binary (Local) inputs]
-* xref:using-secrets-during-build[Input secrets]
+* xref:using-secrets-during-build[Input secrets and config maps]
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config maps/ConfigMaps

File: dev_guide/builds/build_inputs.adoc [hunk 10]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
 executing the build.
 
 [[using-secrets-during-build]]
-== Input Secrets
+== Input Secrets and Config Maps
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/Config Map/ConfigMaps

File: dev_guide/builds/build_inputs.adoc [hunk 11]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
+To add an input secret and/or config map to an existing `BuildConfig`:
+
+. Create the config map, if it does not exist:
++
+----
+$ oc create configmap settings-mvn \
+    --from-file=settings.xml=<path/to/settings.xml>
+----
++
+This creates a new config map named *_settings-mvn_*, which contains the
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config map/ConfigMap

File: dev_guide/builds/build_inputs.adoc [hunk 12]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
+    --from-file=id_rsa=<path/to/.ssh/id_rsa>
 ----
 +
-This creates a new secret named *_secret-npmrc_*, which contains the base64
-encoded content of the *_~/.npmrc_* file.
+This creates a new secret named *_secret-mvn_*, which contains the base64
+encoded content of the *_id_rsa_* private key.
 
-. Add the secret to the `source` section in the existing `BuildConfig`:
+. Add the configMap and secret to the `source` section in the existing
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/configMap/ConfigMap

File: dev_guide/builds/build_inputs.adoc [hunk 13]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
+    - configMap:
+        name: settings-mvn
   secrets:
     - secret:
-        name: secret-npmrc
+        name: secret-mvn
 ----
 
-To include the secret in a new `BuildConfig`, run the following command:
+To include the secret and configMap in a new `BuildConfig`, run the following
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/configMap/ConfigMap

File: dev_guide/builds/build_inputs.adoc [hunk 14]
```diff
@@ -830,10 +859,12 @@ created during the copy process.
...
-Currently, any files with these secrets are world-writable (have `0666`
-permissions) and will be truncated to size zero after executing the *_assemble_*
-script. This means that the secret files will exist in the resulting image, but
-they will be empty for security reasons.
+Input secrets are added as world-writable (have `0666` permissions) and will
+be truncated to size zero after executing the *_assemble_* script. This means 
+that the secret files will exist in the resulting image, but they will be empty 
+for security reasons.
+
+Input config maps are not truncated after the *_assemble_* script completes.
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config maps/ConfigMaps

File: dev_guide/builds/build_inputs.adoc [hunk 15]
```diff
@@ -853,22 +884,22 @@ into that directory, relative to your *_Dockerfile_* location. This makes the
...
-COPY ./secret2 /
+COPY ./secret-dir /secrets
+COPY ./config /
 
-# Create a shell script that will output secrets when the image is run
-RUN echo '#!/bin/sh' > /secret_report.sh
-RUN echo '(test -f /secrets/secret1 && echo -n "secret1=" && cat /secrets/secret1)' >> /secret_report.sh
-RUN echo '(test -f /secret2 && echo -n "relative-secret2=" && cat /secret2)' >> /secret_report.sh
-RUN chmod 755 /secret_report.sh
+# Create a shell script that will output secrets and config maps when the image is run
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config maps/ConfigMaps

File: dev_guide/builds/build_inputs.adoc [hunk 16]
```diff
@@ -883,11 +914,11 @@ they were added. This removal should be part of the *_Dockerfile_* itself.
 [[using-secrets-custom-strategy]]
 === Custom Strategy
 
-When using a `Custom` strategy, all the defined input secrets are available
-inside the builder container in the *_/var/run/secrets/openshift.io/build_*
-directory. The custom build image is responsible for using these secrets
-appropriately. The `Custom` strategy also allows secrets to be defined as
-described in xref:build_strategies.adoc#custom-secrets[Custom Strategy Options].
+When using a `Custom` strategy, all the defined input secrets and config maps
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config maps/ConfigMaps

File: dev_guide/builds/build_inputs.adoc [hunk 17]
```diff
@@ -883,11 +914,11 @@ they were added. This removal should be part of the *_Dockerfile_* itself.
...
 
-When using a `Custom` strategy, all the defined input secrets are available
-inside the builder container in the *_/var/run/secrets/openshift.io/build_*
-directory. The custom build image is responsible for using these secrets
-appropriately. The `Custom` strategy also allows secrets to be defined as
-described in xref:build_strategies.adoc#custom-secrets[Custom Strategy Options].
+When using a `Custom` strategy, all the defined input secrets and config maps 
+are available inside the builder container in the *_/var/run/secrets/openshift.io/build_*
+directory. The custom build image is responsible for using these secrets 
+and config maps appropriately. The `Custom` strategy also allows secrets to be
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/config maps/ConfigMaps

**Comment by adambkaplan** [comment]
@bmcelvee should we also /s/secret/Secret/ when we are referencing a kube Secret object?

**Comment by bmcelvee** [comment]
@adambkaplan I think it's good as is. We tend to keep it as 'secret' (the Kube docs do as well).

File: dev_guide/builds/build_inputs.adoc [hunk 18]
```diff
@@ -731,86 +731,115 @@ archive, and, if valid, the builder changes into that subdirectory before
...
+data to access dependent resources, but it is undesirable for that information 
+to be placed in source control. You can define _input secrets_ and _input 
+ConfigMaps_ for this purpose.
 
-For example, when building a Node.js application, you can set up your private
-mirror for Node.js modules. In order to download modules from that private
-mirror, you have to supply a custom *_.npmrc_* file for the build that contains
-a URL, user name, and password. For security reasons, you do not want to expose
-your credentials in the application image.
+For example, when building an Java application via Maven, you can set up a
```
Comment (bmcelvee) [inline dev_guide/builds/build_inputs.adoc]: s/an Java application via/a Java application with

**Comment by bmcelvee** [comment]
@adambkaplan overall LGTM! Just one more edit and squash commits, then we'll be ready to merge. :zap:

**Comment by adambkaplan** [comment]
@bmcelvee squashed and rebased

**Pull request merged**